File: backend/src/main/java/com/prept/tracker/service/RoadmapService.java
package com.prept.tracker.service;

import com.prept.tracker.domain.entity.Roadmap;
import com.prept.tracker.domain.entity.RoadmapItem;
import com.prept.tracker.domain.entity.User;
import com.prept.tracker.dto.common.PageResponse;
import com.prept.tracker.dto.request.RoadmapItemRequest;
import com.prept.tracker.dto.request.RoadmapRequest;
import com.prept.tracker.dto.response.RoadmapResponse;
import com.prept.tracker.exception.EntityNotFoundException;
import com.prept.tracker.mapper.RoadmapMapper;
import com.prept.tracker.repository.RoadmapItemRepository;
import com.prept.tracker.repository.RoadmapRepository;
import com.prept.tracker.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Service
@RequiredArgsConstructor
public class RoadmapService {

    private final RoadmapRepository roadmapRepository;
    private final RoadmapItemRepository roadmapItemRepository;
    private final RoadmapMapper roadmapMapper;
    private final UserRepository userRepository;

    @Transactional
    public RoadmapResponse create(Long userId, RoadmapRequest request) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new EntityNotFoundException("User not found"));
        Roadmap roadmap = roadmapMapper.toEntity(request);
        roadmap.setUser(user);
        roadmap.setItems(new ArrayList<>());
        Roadmap savedRoadmap = roadmapRepository.save(roadmap);
        if (request.getItems() != null) {
            for (RoadmapItemRequest itemRequest : request.getItems()) {
                RoadmapItem item = roadmapMapper.toItemEntity(itemRequest, savedRoadmap);
                savedRoadmap.getItems().add(item);
            }
        }
        Roadmap saved = roadmapRepository.save(savedRoadmap);
        return roadmapMapper.toResponse(saved);
    }

    @Transactional
    public RoadmapResponse update(Long userId, Long id, RoadmapRequest request) {
        Roadmap roadmap = roadmapRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Roadmap not found"));
        if (!roadmap.getUser().getId().equals(userId)) {
            throw new EntityNotFoundException("Roadmap not found");
        }
        roadmapMapper.updateEntity(request, roadmap);
        roadmap.getItems().clear();
        if (request.getItems() != null) {
            for (RoadmapItemRequest itemRequest : request.getItems()) {
                RoadmapItem item = roadmapMapper.toItemEntity(itemRequest, roadmap);
                roadmap.getItems().add(item);
            }
        }
        Roadmap saved = roadmapRepository.save(roadmap);
        return roadmapMapper.toResponse(saved);
    }

    @Transactional
    public void delete(Long userId, Long id) {
        Roadmap roadmap = roadmapRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Roadmap not found"));
        if (!roadmap.getUser().getId().equals(userId)) {
            throw new EntityNotFoundException("Roadmap not found");
        }
        roadmapRepository.delete(roadmap);
    }

    public RoadmapResponse getById(Long userId, Long id) {
        Roadmap roadmap = roadmapRepository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Roadmap not found"));
        if (!roadmap.getUser().getId().equals(userId)) {
            throw new EntityNotFoundException("Roadmap not found");
        }
        return roadmapMapper.toResponse(roadmap);
    }

    public PageResponse<RoadmapResponse> getByUser(Long userId, Pageable pageable) {
        Page<Roadmap> page = roadmapRepository.findByUserId(userId, pageable);
        return toPageResponse(page);
    }

    private PageResponse<RoadmapResponse> toPageResponse(Page<Roadmap> page) {
        return PageResponse.<RoadmapResponse>builder()
                .content(page.getContent().stream().map(roadmapMapper::toResponse).toList())
                .pageNumber(page.getNumber())
                .pageSize(page.getSize())
                .totalElements(page.getTotalElements())
                .totalPages(page.getTotalPages())
                .last(page.isLast())
                .build();
    }
}

File: frontend/src/services/roadmapService.ts
import api from './api';
import { Roadmap, RoadmapItem, RoadmapCreateRequest, RoadmapUpdateRequest, RoadmapItemCreateRequest, RoadmapItemUpdateRequest } from '../types/roadmap';
import { PageResponse } from '../types/common';

export const roadmapService = {
  getAll: async (params?: { page?: number; size?: number }): Promise<PageResponse<Roadmap>> => {
    const response = await api.get<PageResponse<Roadmap>>('/roadmaps', { params });
    return response.data;
  },
  getById: async (id: number): Promise<Roadmap> => {
    const response = await api.get<Roadmap>(`/roadmaps/${id}`);
    return response.data;
  },
  create: async (data: RoadmapCreateRequest): Promise<Roadmap> => {
    const response = await api.post<Roadmap>('/roadmaps', data);
    return response.data;
  },
  update: async (id: number, data: RoadmapUpdateRequest): Promise<Roadmap> => {
    const response = await api.put<Roadmap>(`/roadmaps/${id}`, data);
    return response.data;
  },
  delete: async (id: number): Promise<void> => {
    await api.delete(`/roadmaps/${id}`);
  },
  getItems: async (roadmapId: number): Promise<RoadmapItem[]> => {
    const response = await api.get<RoadmapItem[]>(`/roadmaps/${roadmapId}/items`);
    return response.data;
  },
  addItem: async (roadmapId: number, data: RoadmapItemCreateRequest): Promise<RoadmapItem> => {
    const response = await api.post<RoadmapItem>(`/roadmaps/${roadmapId}/items`, data);
    return response.data;
  },
  updateItem: async (roadmapId: number, itemId: number, data: RoadmapItemUpdateRequest): Promise<RoadmapItem> => {
    const response = await api.put<RoadmapItem>(`/roadmaps/${roadmapId}/items/${itemId}`, data);
    return response.data;
  },
  deleteItem: async (roadmapId: number, itemId: number): Promise<void> => {
    await api.delete(`/roadmaps/${roadmapId}/items/${itemId}`);
  },
};
export default roadmapService;